File: javascript/mqttRoundTripTime.js
const { PerformanceObserver, performance } = require('perf_hooks');
var mqtt = require('mqtt')
const payloadModule = require('./payloads.js')
// observer for measuring times
const obs = new PerformanceObserver((list) => {
  console.log(list.getEntries()[0].name + ";" + list.getEntries()[0].duration + ";" + list.getEntries()[0].startTime);
});
obs.observe({ entryTypes: ['function', 'measure'] });

// configuration
const host="localhost"
const port=1883
const payload = payloadModule.big;

performance.mark('connecting');
var client  = mqtt.connect('mqtt://localhost:1883');
performance.mark('connected');
performance.measure('connecting to connected', 'connecting', 'connected');

// client events
client.on('connect', function () {

})

// This is a workaround as for some reasen performance.mark does not work in client.on('message')
let lastPublishTime = 0;
client.on('message', function (topic, message) {
  let now = performance.now();
  console.log("published to received" + ";" + (now-lastPublishTime).toString() + ";" + now );
})

// subscribe containing time measurement
function subscribe(){
  performance.mark('subscribing');
  client.subscribe('payload/empty', function (err) {});
  performance.mark('subscribed');
  performance.measure('subscribing to subscribed', 'subscribing', 'subscribed');
}

// subscribe containing time measurement
function publish(){
  performance.mark('publishing');

  client.publish('payload/empty', payload);
  performance.mark('published');
  lastPublishTime = performance.now();
  performance.measure('publishing to published', 'publishing', 'published');
}

// Subscribe to a single topic and publish a msg
subscribe();
publish();


// cleanup
obs.disconnect();
client.end()

File: python/mqttRoundTripTime.py
import paho.mqtt.client as mqtt
import sys
import performanceObserver
import time
import payloads

observer = performanceObserver.PerformanceObserver()

# config
host="localhost"
port=1883
numberOfPublishes = 1
payload = payloads.big

# The callback for CONNACK
def on_connect(client, userdata, flags, rc):
    pass

# The callback for message receive
def on_message(client, userdata, msg):
    observer.mark("received")
    observer.measure("published to received","published", "received")

# cleanup on close
def on_socket_close(client, userdata, sock):
    print("connection closed")
    sys.exit([0])

# create client
client = mqtt.Client()
client.on_connect = on_connect
client.on_message = on_message


# connection is created synchronously
observer.mark("connecting")
client.connect(host, port, 60)
observer.mark("connected")
observer.measure("connecting to connected","connecting", "connected")

# client is watching for events asynchronously
client.loop_start()

# Subscribe to the required Topic
observer.mark("subscribing")
client.subscribe("payload/empty")
observer.mark("subscribed")
observer.measure("subscribing to subscribed","subscribing", "subscribed")

# Publish and measure Round Trip Time
for i in range(0,numberOfPublishes):
    observer.mark("publishing")
    client.publish("payload/empty", payload="", qos=0, retain=False)
    observer.mark("published")
    observer.measure("publishing to published","publishing", "published")
    time.sleep(0.1)

observer.printResult()

# cleanup
client.loop_stop()
sys.exit()
